crates/lib3h/src/message_encoding.rs

```diff
@@ -0,0 +1,339 @@
+    }
+
+    fn handle_msg_from_parent(
+        &mut self,
+        mut msg: MessageEncodingMessageFromParent,
+    ) -> Lib3hResult<()> {
+        //error!("{:?}", msg.backtrace());
+        match msg.take_message().expect("exists") {
+            RequestToChild::Decode { payload } => self.handle_decode(msg, payload),
+            RequestToChild::EncodeHandshake { space_address, id } => {
+                self.handle_encode_handshake(msg, space_address, id)
+            }
+            RequestToChild::EncodePayload { payload } => self.handle_encode_payload(msg, payload),
+        }
+    }
+
+    fn handle_decode(
+        &mut self,
+        msg: MessageEncodingMessageFromParent,
+        payload: Opaque,
+    ) -> Lib3hResult<()> {
+        let payload = match InterimEncodingProtocol::from_slice(&payload) {
+            InterimEncodingProtocol::Handshake {
+                magic,
+                network_id,
+                id,
+            } => {
+                if magic != 0x1f6c {
+                    msg.respond(Err("bad magic".into()))?;
+                    return Ok(());
+                }
+                DecodeData::Handshake {
+                    space_address: network_id,
+                    id,
+                }
+            }
+            InterimEncodingProtocol::Payload { payload } => DecodeData::Payload { payload },
+        };
+        msg.respond(Ok(RequestToChildResponse::DecodeResult { result: payload }))?;
+        Ok(())
+    }
+
+    fn handle_encode_handshake(
+        &mut self,
+        msg: MessageEncodingMessageFromParent,
+        space_address: String,
+        id: String,
+    ) -> Lib3hResult<()> {
+        let payload = InterimEncodingProtocol::Handshake {
+            magic: 0x1f6c,
+            network_id: space_address,
+            id: id,
+        }
+        .to_opaque();
+        msg.respond(Ok(RequestToChildResponse::EncodeHandshakeResult {
+            payload: payload,
+        }))?;
+        Ok(())
+    }
+
+    fn handle_encode_payload(
+        &mut self,
+        msg: MessageEncodingMessageFromParent,
+        payload: Opaque,
+    ) -> Lib3hResult<()> {
+        let payload = InterimEncodingProtocol::Payload { payload }.to_opaque();
+        msg.respond(Ok(RequestToChildResponse::EncodePayloadResult { payload }))?;
+        Ok(())
+    }
+}
+
+impl
+    GhostActor<
+        RequestToParent,
+        RequestToParentResponse,
+        RequestToChild,
+        RequestToChildResponse,
+        Lib3hError,
+    > for MessageEncoding
+{
+    fn take_parent_endpoint(&mut self) -> Option<MessageEncodingParentEndpoint> {
+        std::mem::replace(&mut self.endpoint_parent, None)
+    }
+
+    fn process_concrete(&mut self) -> GhostResult<WorkWasDone> {
+        detach_run!(&mut self.endpoint_self, |es| es.process(self))?;
+        for msg in self.endpoint_self.as_mut().drain_messages() {
+            self.handle_msg_from_parent(msg).expect("no ghost errors");
+        }
+        Ok(false.into())
```

Comment:
Is it worth honoring the `DidWork` contract here? Should be easily implementable.